### PR TITLE
Fix shutting down a notebook when no session is running

### PR DESCRIPTION
**Pull request by seeM**
Addresses #4043.

### QA Notes

I don't know if we'll be able to repro this one, so we might just have to review the code changes.

### CODE REVIEW

File: extensions/positron-notebook-controllers/src/notebookSessionService.ts [hunk 1]
```diff
@@ -252,16 +251,16 @@ export class NotebookSessionService implements vscode.Disposable {
 					session = await startingSessionPromise;
 				} catch (err) {
 					log.error(`Waiting for notebook runtime to start before shutting down failed. Reason ${err}`);
-					throw err;
+
+					// If the session failed to start, we don't need to do anything.
+					return;
 				}
+			} else {
```
Comment (nstrayer) [inline extensions/positron-notebook-controllers/src/notebookSessionService.ts:258]: I find this conditional logic a tiny bit hard to follow. An alternative would be adding a new private method that encapsulates the logic of trying to get the session. E.g. 
```ts
private async _getExistingOrPendingSession(notebookUri: vscode.Uri): Promise<positron.LanguageRuntimeSession | undefined> {
	// Check for an active session first
	const activeSession = this._notebookSessionsByNotebookUri.get(notebookUri);
	if (activeSession) {
		return activeSession;
	}

	// Check for a pending session
	const pendingSession = this._startingSessionsByNotebookUri.get(notebookUri) ||
		this._restartingSessionsByNotebookUri.get(notebookUri);

	if (!pendingSession) {
		return undefined;
	}

	try {
		return await pendingSession;
	} catch (err) {
		log.error(`Waiting for notebook runtime to start before shutting down failed. Reason ${err}`);
		return undefined;
	}
}
```
So then it could be:
```ts
async doShutdownRuntimeSession(notebookUri: vscode.Uri): Promise<void> {
	const session = await this._getExistingOrPendingSession(notebookUri);
	if (!session) {
		// No active or pending session found, nothing to shut down
		return;
	}
	...
}
```

The logic though is sound so it's not terribly important!

File: extensions/positron-notebook-controllers/src/notebookSessionService.ts [hunk 2]
```diff
@@ -252,16 +251,16 @@ export class NotebookSessionService implements vscode.Disposable {
...
+				// If there's no session and no starting session, we don't need to do anything.
+				return;
 			}
 		}
 
-		// Ensure that we have a session.
-		if (!session) {
-			throw new Error(`Tried to shutdown runtime for notebook without a running runtime: ${notebookUri.path}`);
```
Comment (nstrayer) [inline extensions/positron-notebook-controllers/src/notebookSessionService.ts:262]: Should we still spit something out into a log here? Or is the idea that it's not going to be possible to get to this situation anymore?

**Review by nstrayer — DISMISSED** [review]
LGTM. Just an organization idea.

File: extensions/positron-notebook-controllers/src/notebookSessionService.ts (hunk 2 above)
Comment (seeM) [reply extensions/positron-notebook-controllers/src/notebookSessionService.ts:262]: Yeah, I'll add a debug log

File: extensions/positron-notebook-controllers/src/notebookSessionService.ts (hunk 1 above)
Comment (seeM) [reply extensions/positron-notebook-controllers/src/notebookSessionService.ts]: I like your suggestion! Done in 89b777504d2
